lib/chef/provider/package.rb

```diff
@@ -439,16 +467,19 @@ def target_version_array
             each_package do |package_name, new_version, current_version, candidate_version|
               case action
               when :upgrade
-                if target_version_already_installed?(current_version, new_version)
+                if version_equals?(current_version, new_version)
                   # this is an odd use case
                   Chef::Log.debug("#{new_resource} #{package_name} #{new_version} is already installed -- you are equality pinning with an :upgrade action, this may be deprecated in the future")
                   target_version_array.push(nil)
-                elsif target_version_already_installed?(current_version, candidate_version)
+                elsif version_equals?(current_version, candidate_version)
                   Chef::Log.debug("#{new_resource} #{package_name} #{candidate_version} is already installed")
                   target_version_array.push(nil)
                 elsif candidate_version.nil?
                   Chef::Log.debug("#{new_resource} #{package_name} has no candidate_version to upgrade to")
                   target_version_array.push(nil)
+                elsif version_compare(current_version, candidate_version) == 1 && !new_resource.allow_downgrade
```

Comment:
Based on my limited understanding of the package machinery, I think this will catch an explicit version too? If so that would be a potentially undesirable behavior change. I very well may be misreading things though.

Reply:
It would be simple enough to skip this step if new_resource.version has been specified, which would keep behaviour of preferring an explicit version over the action - sound good?

Reply:
I think the current behavior if you specify a version would be to crash because you can't downgrade, but I'd have to check...

Reply:
So having tested a sample size of two (yum and dnf), we don't behave consistently here. DNF will see things as up to date and do nothing, yum will act as if there's no current version and upgrade from uninstalled -> specified version. The question is whether we want to have all providers behave as they currently do (ie not consistently), or standardise this somewhat..but that then brings us to @coderanger's original point that this might be a compat change.